src/celeritas/global/alongstep/AlongStepRZMapFieldMscAction.cc

```diff
@@ -75,15 +78,41 @@ AlongStepRZMapFieldMscAction::AlongStepRZMapFieldMscAction(
 void AlongStepRZMapFieldMscAction::execute(CoreParams const& params,
                                            CoreStateHost& state) const
 {
-    auto execute = make_along_step_track_executor(
-        params.ptr<MemSpace::native>(),
-        state.ptr(),
-        this->action_id(),
-        AlongStep{UrbanMsc{msc_->ref<MemSpace::native>()},
-                  detail::RZMapFieldPropagatorFactory{
-                      field_->ref<MemSpace::native>()},
-                  detail::FluctELoss{fluct_->ref<MemSpace::native>()}});
-    return launch_action(*this, params, state, execute);
+    using namespace ::celeritas::detail;
+
+    auto launch_impl = [&](auto&& execute_track) {
+        return launch_action(
+            *this,
+            params,
+            state,
+            make_along_step_track_executor(
+                params.ptr<MemSpace::native>(),
+                state.ptr(),
+                this->action_id(),
+                std::forward<decltype(execute_track)>(execute_track)));
+    };
+
+    if (this->has_msc())
+    {
+        launch_impl(
+            MscStepLimitApplier{UrbanMsc{msc_->ref<MemSpace::native>()}});
+    }
+    launch_impl(PropagationApplier{
+        RZMapFieldPropagatorFactory{field_->ref<MemSpace::native>()}});
+    if (this->has_msc())
+    {
+        launch_impl(MscApplier{UrbanMsc{msc_->ref<MemSpace::native>()}});
+    }
+    launch_impl(detail::TimeUpdater{});
+    if (this->has_fluct())
+    {
+        launch_impl(ElossApplier{FluctELoss{fluct_->ref<MemSpace::native>()}});
+    }
+    else
+    {
+        launch_impl(ElossApplier{MeanELoss{}});
+    }
+    launch_impl(TrackUpdater{});
```

Comment:
It would be interesting to see if there's any overhead from launching a bunch of separate CPU "kernels" versus a single one. The latter only has to check once for a track being disabled, for example. It might be better to put all the conditionals on the inside of the loop. There's going to be a tradeoff between improved locality of state data and locality of params data.

Reply:
That's a good point, I can look into that.

Reply:
Looking at a couple problems (one using general linear and one using uniform msc along-step) it looks like using a single CPU "kernel" results in ~5-10% improvement, so we might want to consider converting all of these.